theatre/studio/src/uiComponents/usePresence.tsx

```diff
@@ -0,0 +1,231 @@
+  /**  */
+  addRelationships(
+    itemKey: StudioSheetItemKey,
+    config: Array<RelationshipConfig>,
+  ): () => void
+}
+
+const presenceCtx = React.createContext<Presence>(createPresence())
+export function ProvidePresenceRoot({
+  presence = createPresence(),
+  children,
+}: React.PropsWithChildren<{presence?: Presence}>) {
+  return React.createElement(
+    presenceCtx.Provider,
+    {children, value: presence},
+    children,
+  )
+}
+
+const PRESENCE_ITEM_DATA_ATTR = 'data-pi'
+// const PRESENCE_ITEM_DEBUG_DATA_ATTR = 'data-pi-loc'
+
+export default function usePresence(config: {
+  key: StudioSheetItemKey | undefined
+}): [
+  props: {[attr: `data-${string}`]: string},
+  focus: {
+    current: Hovered | undefined
+    useRelationships(
+      getRelationships: () => Array<RelationshipConfig>,
+      // QUESTION: is this right?
+      deps: any[],
+    ): void
+  },
+] {
+  const presence = React.useContext(presenceCtx)
+  const focus = presence.useFocused(config.key)
+
+  return [
+    {
+      [PRESENCE_ITEM_DATA_ATTR]: config.key as string,
+      // [PRESENCE_ITEM_DEBUG_DATA_ATTR]: config.areaDebugName,
+    },
+    {
+      current: focus,
+      useRelationships(getRelationships, deps) {
+        useEffect(() => {
+          console.log('deps', deps.length, ...deps)
+          return (
+            config.key &&
+            presence.addRelationships(config.key, getRelationships())
+          )
+        }, [config.key, ...deps])
+      },
+    },
+  ]
+}
+
+export function usePresenceListeners(target: HTMLElement | null | undefined) {
+  const presence = React.useContext(presenceCtx)
+  const logger = useLogger('PresenceListeners')
+  useEffect(() => {
+    let cleanup: (() => void) | undefined = undefined
+    let last: any
+    if (!target) return
+    const onMouseOver = (event: MouseEvent) => {
+      if (event.target instanceof Node) {
+        const found = closest(event.target, `[${PRESENCE_ITEM_DATA_ATTR}]`)
+        if (found) {
+          const itemKey = found.getAttribute(PRESENCE_ITEM_DATA_ATTR)
+          if (last !== itemKey) {
+            last = itemKey
+            presence.setUserHover(itemKey as StudioSheetItemKey | undefined)
+            logger._debug('Updated current hover', {itemKey})
+          }
+          return
+        }
+
+        // remove hover
+        if (last != null) {
+          last = null
+          presence.setUserHover(undefined)
+          logger._debug('Cleared current hover')
+        }
+      }
+    }
+
+    target.addEventListener('mouseover', onMouseOver)
+
+    return () => {
+      cleanup?.()
+      target.removeEventListener('mouseover', onMouseOver)
+      // remove hover
+      if (last != null) {
+        last = null
+        logger._debug('Cleared current hover as part of cleanup')
+      }
+    }
+  }, [target, presence])
+}
+
+/**
+ * Traverse upwards from the current element to find the first element that matches the selector.
+ */
+function closest(
```

Comment:
nit: rename `closest` to `selectClosestHTMLAncestor`

Reply:
Good suggestion!

Reply:
Applied